ambry-account/src/main/java/com/github/ambry/account/MySqlAccountService.java

```diff
@@ -55,288 +47,141 @@ public class MySqlAccountService extends AbstractAccountService {
+  public MySqlAccountService(AccountServiceMetricsWrapper accountServiceMetrics, MySqlAccountServiceConfig config,
       MySqlAccountStoreFactory mySqlAccountStoreFactory, Notifier<String> notifier) throws SQLException, IOException {
     super(config, Objects.requireNonNull(accountServiceMetrics, "accountServiceMetrics cannot be null"), notifier);
     this.config = config;
-    this.mySqlAccountStoreFactory = mySqlAccountStoreFactory;
-    try {
-      this.mySqlAccountStore = mySqlAccountStoreFactory.getMySqlAccountStore();
-    } catch (SQLException e) {
-      logger.error("MySQL account store creation failed", e);
-      // If it is a non-transient error like credential issue, creation should fail.
-      // Otherwise, continue account service creation and initialize cache with metadata from local file copy
-      // to serve read requests. Connection to MySql DB will be retried during periodic sync. Until then, write
-      // requests will be blocked.
-      if (MySqlDataAccessor.isCredentialError(e)) {
-        // Fatal error, fail fast
-        throw e;
+    Supplier<MySqlAccountStore> accountStoreSupplier = () -> {
+      try {
+        this.mySqlAccountStore = mySqlAccountStoreFactory.getMySqlAccountStore();
+        return this.mySqlAccountStore;
+      } catch (SQLException e) {
+        logger.error("MySQL account store creation failed", e);
+        // If it is a non-transient error like credential issue, creation should fail.
+        // Otherwise, continue account service creation and initialize cache with metadata from local file copy
+        // to serve read requests. Connection to MySql DB will be retried during periodic sync. Until then, write
+        // requests will be blocked.
+        if (MySqlDataAccessor.isCredentialError(e)) {
+          // Fatal error, fail fast
+          throw new RuntimeException("Failed to create account store due to credential issue", e);
+        } else {
+          return null;
+        }
       }
-    }
-    accountServiceMetrics.trackTimeSinceLastSync(this::getTimeInSecondsSinceLastSync);
-    accountServiceMetrics.trackContainerCount(this::getContainerCount);
+    };
+    Supplier<MySqlAccountStore> accountStoreSupplierNew = () -> {
+      try {
+        this.mySqlAccountStore = mySqlAccountStoreFactory.getMySqlAccountStoreNew();
+        return this.mySqlAccountStore;
+      } catch (SQLException e) {
+        logger.error("MySQL account store creation failed", e);
+        if (MySqlDataAccessor.isCredentialError(e)) {
+          throw new RuntimeException("Failed to create account store due to credential issue", e);
+        } else {
+          return null;
+        }
+      }
+    };
+    accountServiceMetrics.trackTimeSinceLastSync(this::getTimeInSecondsSinceLastSync,
+        this::getTimeInSecondsSinceLastSyncNew);
+    accountServiceMetrics.trackContainerCount(this::getContainerCount, this::getContainerCountNew);
```

Comment:
nit:
`
@FunctionalInterface
public interface ThrowingSupplier<T, E> {
    T get() throws E;
}

private Supplier<MySqlAccountStore> callSupplierWithException(ThrowingSupplier<MySqlAccountStore, SQLException> ts) {
    return () -> {
        try {
            return ts.get();
        } catch (SQLException e) {
           logger.error("MySQL account store creation failed", e);
            if (MySqlDataAccessor.isCredentialError(e)) {
                throw new RuntimeException("Failed to create account store due to credential issue", e);
            } else {
                return null;
            }
        }
    };
}

cachedAccountService = new CachedAccountService(callSupplierWithException(factory::getAccountStore), accountServiceMetrics.getAccountServiceMetrics(), config, notifier, config.backupDir, scheduler);
`

Reply:
Thanks, updated.